.circleci/real_config.yml

```
@@ -2700,6 +2700,87 @@ jobs:
           mentions: <<parameters.slack-mentions>>
           channel: <<parameters.slack-channel>>
 
+  run-shared-cluster-cleanup:
+      parameters:
+        cluster-id:
+          type: string
+          default: ${GKE_CLUSTER_NAME}
+        region:
+          type: string
+          default: ${GKE_REGION}
+        gcloud-service-key:
+          default: GCLOUD_SERVICE_KEY
+          description: The gcloud service key
+          type: env_var_name
+        google-compute-zone:
+          default: GOOGLE_COMPUTE_ZONE
+          description: The Google compute zone to connect with via the gcloud CLI
+          type: env_var_name
+        google-project-id:
+          default: GOOGLE_PROJECT_ID
+          description: The Google project ID to connect with via the gcloud CLI
+          type: env_var_name
+      circleci_ip_ranges: true
+      docker:
+        - image: <<pipeline.parameters.docker-image>>
+      steps:
+        - set-cluster-id:
+              cluster-id: <<parameters.cluster-id>>
+        - gcloud/install:
+              version: "412.0.0"
+        - kubernetes/install-kubectl
+        - gcloud/initialize:
+            gcloud-service-key: <<parameters.gcloud-service-key>>
+            google-compute-zone: <<parameters.google-compute-zone>>
+            google-project-id: <<parameters.google-project-id>>
+        - run:
+            command: |
+              tries=5
+              until gcloud components install gke-gcloud-auth-plugin --quiet; do
+                if [[ $((--tries)) -eq 0 ]]; then
+                  exit 1
+                fi
+                sleep 15
+              done
+              echo "export USE_GKE_GCLOUD_AUTH_PLUGIN=True" >> $BASH_ENV
+            name: Install GKE auth plugin
+        - run:
+            name: Get Kubeconfig
+            command: gcloud container clusters get-credentials ${CLUSTER_ID} --project ${GOOGLE_PROJECT_ID} --region <<parameters.region>>
+        - run:
+            name: Delete GKE CI Cluster Namespaces
+            command: |
+              kubectl get namespace | grep -Eo "^test-cpu-[a-z0-9]+-[a-z0-9]+-[0-9]" | xargs -L1 kubectl delete namespace || true
+        - run:
+            name: Delete GCS CI Buckets
+            command: |
+              gsutil ls -p ${GOOGLE_PROJECT_ID} | grep -Eo "^gs://test-cpu-[a-z0-9]+-[a-z0-9]+-[0-9]-bucket" | xargs -L1 gsutil rm -r || true
+        - run:
+            name: Delete Firewall Rules
+            command: |
+              gcloud compute networks get-effective-firewalls $GCP_NETWORK_NAME --project "$GCP_PROJECT_ID" \
+                --format="table(name)" | tail -n +2 | \
+                while read fw; do
+                  if [[ $fw =~ "k8s" ]]; then
+                    gcloud compute firewall-rules delete "$fw" --quiet --project "$GCP_PROJECT_ID"
+                  fi
+                done
```

The gcloud delete commands mostly take multiple targets for delete, so this could be done (faster) with a single command along the lines of this:

```suggestion
              gcloud compute firewall-rules delete "$fw" --quiet --project "$GCP_PROJECT_ID" $( \
                gcloud compute networks get-effective-firewalls $GCP_NETWORK_NAME --project "$GCP_PROJECT_ID" --regexp=".*k8s.*" \
                --format="table(name)" | tail -n +2
                )
```

Piping into a while loop like that means any command inside the loop which reads from stdin will consume output before the `read` gets another chance, so it's generally ideal to avoid that structure for maintenance (works now, but what if someone adds a second command in the loop later?).

---

This would actually be even easier with a cloud-custodian yaml file which could just select the stuff to remove. :D

---

It seems like the `kubectl delete namespace` command removes firewall rules, target pools, and forwarding rules (since those resources are namespaced)! So, I'll remove these three steps from the cleaup job

---

Awesome, even better!